### PR TITLE
:recycle: [#2257] Fetch subscriptions for user directly from Laposta

### PR DESCRIPTION
**Pull request by stevenbal**
task: https://taiga.maykinmedia.nl/project/open-inwoner/task/2257

Notes:
- Unfortunately it is not possible to retrieve all subscriptions for a given email adress in a single API call, so we have to perform an API call for each list
- The API applies throttling based on the used API key, limiting to 30 calls per minute (or 120 if it's a paid account), so I decided to apply some caching to not let the number of request too out of hand

### CODE REVIEW

**Review by pi-sigma — APPROVED** [review]

File: src/open_inwoner/laposta/client.py [hunk 1]
```diff
@@ -71,6 +88,26 @@ def remove_subscription(self, list_id: str, member_id: str) -> Member | None:
...
+        # The API cannot deal with urlencoded email addresses that contain a `+`,
+        # so instead we fetch all members for a given list and check if there is a
+        # member with a matching email
+        subscribed_to = {}
+        for list_id in list_ids:
+            response = self.get("member", params={"list_id": list_id})
+            if response.status_code == 200:
+                data = get_json_response(response)
+                for member_data in data["data"]:
+                    member = Member(**member_data["member"])
```
Comment (Bartvaderkin) [inline src/open_inwoner/laposta/client.py:106]: The list of all members could be very long: maybe we don't need throwaway instances here but can use the dict.

File: src/open_inwoner/laposta/client.py [hunk 2]
```diff
@@ -53,9 +61,18 @@ def create_subscription(self, list_id: str, user_data: UserData) -> Member | Non
...
 
-    def remove_subscription(self, list_id: str, member_id: str) -> Member | None:
-        response = self.delete(f"member/{member_id}", params={"list_id": list_id})
-
+    def remove_subscription(
+        self, list_id: str, member_id: str, email: str
+    ) -> Member | None:
+        # Ensure the current subscriptions for this email address are fetched again after
+        # this API call
+        cache.delete(f"laposta_list_subscriptions:{email}")
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/open_inwoner/laposta/client.py:69]: You're right, I'll move the cache reset for create/delete to the bottom of the functions

**Review by Bartvaderkin — CHANGES_REQUESTED** [review]

File: src/open_inwoner/laposta/client.py [hunk 3]
```diff
@@ -71,6 +88,26 @@ def remove_subscription(self, list_id: str, member_id: str) -> Member | None:
...
 
+    @cache_result(
+        "laposta_list_subscriptions:{email}", timeout=settings.CACHE_LAPOSTA_API_TIMEOUT
+    )
+    def get_subscriptions_for_email(
+        self, list_ids: list[str], email: str
+    ) -> dict[str, str]:
+        # The API cannot deal with urlencoded email addresses that contain a `+`,
+        # so instead we fetch all members for a given list and check if there is a
+        # member with a matching email
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/open_inwoner/laposta/client.py:99]: I tried it both with and without quoting the email:

`alex+111111110@maykinmedia.nl` and `foo@bar.org` are subscribed to the list:
Screenshot from 2024-04-08 11-17-59

I can directly fetch the relation for `foo@bar.org`:
Screenshot from 2024-04-08 11-19-01

But I cannot do that for the email with a plus:
Screenshot from 2024-04-08 11-18-11

Screenshot from 2024-04-08 11-18-45

File: src/open_inwoner/laposta/forms.py [hunk 4]
```diff
@@ -37,10 +37,11 @@ def __init__(self, user=None, *args, **kwargs):
...
                 ]
 
-            self.fields["newsletters"].initial = [
-                subscription.list_id
-                for subscription in Subscription.objects.filter(user=user)
-            ]
+            subscription_mapping = laposta_client.get_subscriptions_for_email(
+                limited_to, request.user.email
+            )
+            request.session["subscription_mapping"] = subscription_mapping
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/open_inwoner/laposta/forms.py:43]: It's just for this request, so I'll try moving it to an attribute instead :+1:

File: src/open_inwoner/laposta/client.py (hunk 1 above)
Comment (stevenbal) [reply src/open_inwoner/laposta/client.py:106]: The `member_id` is not the same for the same email across different `list_id`s unfortunately, so I would still have to parse the `member_id`

File: src/open_inwoner/laposta/client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bartvaderkin) [reply src/open_inwoner/laposta/client.py:99]: Good tests, but it feels like something is wrong that we need to download the while subscriber database to look at a single user's subscriptions.

[the remark this answers is not in the document]
Comment (stevenbal) [reply src/open_inwoner/laposta/client.py:99]: I'll email Laposta to ask if they know about this issue, but in the meantime I'm not sure if there is another (easier) way to fix this

[the remark this answers is not in the document]
Comment (Bartvaderkin) [reply src/open_inwoner/laposta/client.py:99]: Ok cool, will be interesting. If it takes too long or they don't have anything we can use this solution and see.

[the remark this answers is not in the document]
Comment (alextreme) [reply src/open_inwoner/laposta/client.py:99]: But... this is mentioned in their API docs:

image

[the remark this answers is not in the document]
Comment (stevenbal) [reply src/open_inwoner/laposta/client.py:99]: oh well, I'll do that instead then

[the remark this answers is not in the document]
Comment (stevenbal) [reply src/open_inwoner/laposta/client.py]: Removed the workaround and replaced it with double quoting in https://github.com/maykinmedia/open-inwoner/pull/1138/commits/504b250b3809ade6de8f0f888cc9fc1d7814a6c5

**Review by Bartvaderkin — APPROVED** [review]
Noting I need to update this with the verified_email

**Pull request merged**